### PR TITLE
Add model and guide auxiliary variable elimination to minipyro.elbo

### PR DESCRIPTION
**Pull request by eb8680**
This exploratory PR adds a few lines of code to `minipyro.elbo` that integrate out auxiliary variables in the model and guide, as in `pyro.infer.TraceEnum_ELBO`.

## Tasks
- [x] Add an example or two
- [x] Add tests

## Triaged
- Troubleshoot performance issues
- Switch to using Expectation term #132 
- Mixed sampling and exact integration in `elbo`

### CODE REVIEW

**Comment by fritzo** [comment]
I really like the idea of implementing a generalized `TraceEnum_ELBO`. It would be nice to support sample site annotations like `infer={"enumerate": "parallel"}` or something more modern like `infer={"delayed": True}` or `infer={"exact": True}`. Then we can add this new annotation to Pyro and compare the two `TraceEnum_ELBO` implementations. WDYT?

**Comment by eb8680** [comment]
Yeah, I think that makes sense. That might also let us reuse a bunch of
tests from Pyro.

File: funsor/minipyro.py [hunk 1]
```diff
@@ -71,6 +79,12 @@ def get_param_store():
     return PARAM_STORE
 
 
+def clear_param_store():
```
Comment (fritzo) [inline funsor/minipyro.py]: `PARAM_STORE.clear()`

File: funsor/minipyro.py [hunk 2]
```diff
@@ -212,6 +226,55 @@ def postprocess_message(self, msg):
...
+        if msg["type"] == "sample" and msg["value"] is None:
+            if msg["infer"].get("enumerate", None) == "parallel":
+                msg["value"] = funsor.Variable(msg["name"], msg["fn"].output)
+            elif msg["infer"].get("enumerate", None) is None:
+                msg["value"] = msg["fn"](*msg["args"])  # default to eager
+            else:
+                raise ValueError("{} not a supported enumeration type".format(msg["infer"]["enumerate"]))
+
+
+class infer_config(Messenger):
```
Comment (fritzo) [inline funsor/minipyro.py]: I think we can omit this

File: funsor/minipyro.py [hunk 3]
```diff
@@ -212,6 +226,55 @@ def postprocess_message(self, msg):
...
+
+    def __init__(self, fn, config_fn):
+        self.config_fn = config_fn
+        super(infer_config, self).__init__(fn)
+
+    def process_message(self, msg):
+        msg["infer"].update(self.config_fn(msg))
+
+
+def _config_enumerate(default, num_samples=None):
```
Comment (fritzo) [inline funsor/minipyro.py]: I think we can omit this

File: funsor/minipyro.py [hunk 4]
```diff
@@ -55,6 +55,14 @@ def expand_inputs(self, name, size):
         funsor_dist = self.funsor_dist + funsor.torch.Tensor(torch.zeros(size), inputs)
         return Distribution(funsor_dist)
 
+    @property
+    def has_enumerate_support(self):
```
Comment (fritzo) [inline funsor/minipyro.py]: Omit this, funsor shouln't preserve this legacy notion

File: funsor/minipyro.py [hunk 5]
```diff
@@ -212,6 +226,55 @@ def postprocess_message(self, msg):
...
+            return {"enumerate": site["infer"].get("enumerate", default),
+                    "num_samples": site["infer"].get("num_samples", num_samples)}
+        if getattr(site["fn"], "has_enumerate_support", False):
+            return {"enumerate": site["infer"].get("enumerate", default)}
+        return {}
+
+    return config_fn
+
+
+def config_enumerate(model=None, default="parallel", num_samples=None):
```
Comment (fritzo) [inline funsor/minipyro.py]: delete this

File: funsor/minipyro.py (hunk 1 above)
Comment (fritzo) [inline funsor/minipyro.py]: Delete this; use `pyro.get_param_store().clear()` instead

File: test/test_minipyro.py [hunk 6]
```diff
@@ -237,3 +264,268 @@ def guide():
...
+        elbo = elbo.differentiable_loss if backend == "pyro" else elbo
+        hand_loss = elbo(hand_model, hand_guide, data)
+        _check_loss_and_grads(hand_loss, auto_loss)
+
+
+@pytest.mark.xfail(reason="missing patterns")
+@pytest.mark.parametrize("jit", [False, True], ids=["py", "jit"])
+@pytest.mark.parametrize("exact", [
+    True,
+    xfail_param(False, reason="mixed sampling and exact not implemented yet")
```
Comment (eb8680) [inline test/test_minipyro.py]: I'll defer mixed sampling and exact integration to another PR

**Comment by fritzo** [comment]
Thanks for completing this on short notice @eb8680, hopefully we can get it into prod soon :smile:

File: funsor/minipyro.py [hunk 7]
```diff
@@ -400,31 +401,61 @@ def elbo(model, guide, *args, **kwargs):
...
-    sum_vars = frozenset().union(guide_log_joint.log_factors,
-                                 model_log_joint.log_factors)
     elbo = Expectation(tuple(log_probs),
                        tuple(costs),
                        sum_vars=sum_vars,
                        prod_vars=plates)
 
     loss = -elbo
     assert not loss.inputs
-    assert isinstance(loss, funsor.torch.Tensor), loss.pretty()
```
Comment (fritzo) [inline funsor/minipyro.py]: Hmm, this was super useful for debugging. Mind if I resurrect it?

Comment (fritzo) [reply funsor/minipyro.py]: Aha, we can move it into SVI, where the assumption is needed.

**Pull request merged**